src/librbd/io/ImageRequest.cc

```diff
@@ -342,24 +318,14 @@ int ImageReadRequest<I>::clip_request() {
     return r;
   }
 
-  uint64_t buffer_length = 0;
-  auto &image_extents = this->m_image_extents;
-  for (auto &image_extent : image_extents) {
-    buffer_length += image_extent.second;
-  }
-  this->m_aio_comp->read_result.set_clip_length(buffer_length);
+  util::set_read_clip_length(this->m_image_extents, this->m_aio_comp);
   return 0;
 }
 
 template <typename I>
 bool ImageReadRequest<I>::finish_request_early() {
-  auto total_bytes = this->get_total_length();
-  if (total_bytes == 0) {
-    auto *aio_comp = this->m_aio_comp;
-    aio_comp->set_request_count(0);
-    return true;
-  }
-  return false;
+  return util::finish_request_early_if_nodata(this->m_image_extents,
```

Comment:
Nit: see previous comment about doing this logic in one place only one time

Reply:
Moved to ImageDispatcher.